### ISSUE TITLE
upload file of multiple files of 256M is very slow

### ISSUE THREAD

**Issue by TamarLevy** [tiangolo/fastapi#5413]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import File, UploadFile
import aiofiles

# Upload file using fastapi
@app.post("/upload")
async def upload1(files: UploadFile = File(...)):
    for file in files:
        try:
            contents = await file.read()
            async with aiofiles.open(file.filename, 'wb') as f:
                await f.write(contents)
        except Exception:
            return {"message": "There was an error uploading the file(s)"}
        finally:
            await file.close()

    return {"message": f"Successfuly uploaded {[file.filename for file in files]}"} 

# STREAM method to upload file

from fastapi import Request
import aiofiles

@app.post('/upload')
async def upload2(request: Request):
    try:
        filename = request.headers['filename']
        async with aiofiles.open(filename, 'wb') as f:
            async for chunk in request.stream():
                await f.write(chunk)
    except Exception:
        return {"message": "There was an error uploading the file"}
     
    return {"message": f"Successfuly uploaded {filename}"}
```

### Description

if I try to upload using upload1 in parallel 6 files of 256M each, fastapi upload is stuck(the function upload1 is not even entered until all 6 files are copied), from reading about it here:
https://stackoverflow.com/questions/65342833/fastapi-uploadfile-is-slow-compared-to-flask/70667530#70667530
I understand that if the size is bigger then 1MB then the file is written to disk (to a tmp directory, but I couldn't find it there for some reason)
I tried second approach using upload2 (stream) but then my upload was stuck in this line:
async for chunk in request.stream()
Im wondering if there is anything wrong with these implementations (upload1 or upload2), why in upload1, all the files are uploaded and then upload1 function is entered (for each upload)
how to generate a bigger chunk for upload2?

### Operating System

Linux

### Operating System Details

Vmware Ubuntu 18.4

### FastAPI Version

0.78.0

### Python Version

python 3.8

**Comment by JarroVGIT** [comment]
`Upload1` expects a `File()`, not a list of `File()`s, so why are you looping through it? Have you tried using `UploadFile()` instead? How are you uploading? What do you mean when you say 
> the function upload1 is not even entered until all 6 files are copied

copied from where?

**Comment by TamarLevy** [comment]
my intention was to upload multiple files in parallel to the server to simulate an upload request coming from different clients
so I simulate it with 6 upload of 256M file each, Im not looping through it Im sending the files asynchronically to the server 
so all the requests are sent in parallel.
This is my code for sending 6 files asynchronically:

def v1_upload(upload_url, files):
       res = requests.post(upload_url, files=files)
       return res 

async def request_upload_async(upload_url, file):
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(None, v1_upload,  upload_url, file)

 async def upload_async_multiple(num_async, upload_url, files):
        await asyncio.gather(*(self.request_upload_async(upload_url[i], files[i]) for i in range(num_async)))


loop = asyncio.get_event_loop()
loop.run_until_complete(upload_async_multiple(6, upload_url, files))
elapsed = (time.perf_counter() - start)

the upload1 of 256M of 6 files took approximately 12sec, when I measure one upload1 of 256M it took 2 seconds which means there is no parallelism in the upload1() at all, I tried to debug the upload1() code to see where it spends most of the time and I see that upload1 is not even entered (meaning the first line of code of the function is not even called) until all 6 files received to the server, I understood from reading this link: https://stackoverflow.com/questions/65342833/fastapi-uploadfile-is-slow-compared-to-flask/70667530#70667530
that every file that is uploaded tried to be copied to RAM and if the content of the file is bigger then 1MB then the files are written to a temporary directory, I tried to do watch on my temporary directory but I didnt see any files copied there.
it look like all the files are copied and then upload1() is called for each file.
I was wondering if this is some expected behavior when 6 files are uploaded in parallel or maybe upload1() can be optimized further to handle such a situation.

**Comment by JarroVGIT** [comment]
I think this is more of an OS issue than a programming issue. For example: try to copy-paste one file to another folder on you machine, and while it is copying, do that for several other files as well. The copying time will increase drastically. So, since 6 files are written to disk simultaneously and then again read and written **again**, it does make sense this will take a lot longer for multiple files concurrently (I am having doubt that there is any parallel IO going on at all on the OS level but I could be mistaken). 

So couple of pointers on how to improve this situation: 
- In production environments (like a Kubernetes cluster), you could use multi-write (block-)storage and multiple workers.
- Don't re-read the temp file and then rewrite it somewhere else, but copy it instead.

Maybe I am overlooking something simple though..

**Comment by TamarLevy** [comment]
Hi Thanks for the quick response.
I dont think its an OS thing, its only 6 files and the time it takes me to upload 6 files one after another and the time it take to upload 6 file in parallel is the same, which means there is no improvement when you upload 6 files in parallels.
about the second point of how to improve the code:
how exactly do you do that? (copy the file from temp, instead of reading and then rewriting)

**Comment by JarroVGIT** [comment]
I found the topic interesting and so far I found that SATA drives will not write parallel but nvme drives can.

> the time it takes me to upload 6 files one after another and the time it take to upload 6 file in parallel is the same, which means there is no improvement when you upload 6 files in parallels.

This is exactly my point; if you can’t write it to disk in parallel (due to disk limitations) then how do you expect it to run faster with more files? It would write sequentially and thus result in 6 x the original speed?

**Comment by Kludex** [comment]
How is it with pure Starlette?

**Comment by TamarLevy** [comment]
BTW I have a system with nvme and not SATA
So you're saying that with pure Starlette I can upload a file to the server without the redundancy of writing the file to tmp directory then reading it and writing it to some other file?

**Comment by Kludex** [comment]
No... There's nothing hidden on my question... I'm asking the behavior on Starlette. 🤔 

If the issue still happens with pure Starlette, then it's not a FastAPI issue.

**Comment by TamarLevy** [comment]
Im sorry I didnt understand you earlier.
Yes this behavior still exists, Im not using the earliest fastapi package though (0.78.0), I dont know if that matters.
Thanks for the quick response :-), I will forward my question to starlette

**Comment by Kludex** [comment]
This is probably an issue with the `MultiPartParser` we use. I'm already aware of some performance issues there.

The issue on Starlette is welcome tho. 👍

The version of FastAPI doesn't matter here, we didn't change that part of the code on Starlette for a long time.

**Comment by TamarLevy** [comment]
ok I have another issue, that Im trying to figure out maybe you can help me :-)
I also try to upload data to server as a stream, which I succeed but the default chunk size of the stream is very small (64K) and I want to upload a bigger chunk specially if I try to upload 256MB of data:
@app.post('/upload')
async def upload2(request: Request):
    try:
        filename = request.headers['filename']
        async with aiofiles.open(filename, 'wb') as f:
            async for chunk in request.stream():
                await f.write(chunk)
    except Exception:
        return {"message": "There was an error uploading the file"}
     
    return {"message": f"Successfuly uploaded {filename}"}

someone in stuckoverflow(https://stackoverflow.com/questions/73322065/how-to-increase-the-chunk-size-when-sending-an-http-stream-upload-request-throug) 
told me that you can change the chunk size be doing this:
rand_file = generate_rand_data_file(data_size=file_size)
header = {'filename': os.path.basename(rand_file), 'content-length': str(os.path.getsize(rand_file))}
 def gen_file_stream():
    with open(rand_file , "rb") as f:
        while True:
            data = f.read(1 * 1024 * 1024)
            if not data:
                break
            yield data
res = client.post('/upload', data=gen_file_stream, headers=header)

This code does not work and raise an exception, do you know if its even possible to pass in `data` parameter a function that reads the data in chunks?

I really appreciate your quick responses :-)

**Comment by panla** [comment]
client upload split section

**Comment by tiangolo** [comment]
It seems you are not reading/processing the files concurrently, in your code, you are iterating on each file before you start doing the await stuff.

Try to create an async function that takes a file and processes it, then create a task group and add a task for each file.

Read the docs in Asyncer: https://asyncer.tiangolo.com/

**Comment by TamarLevy** [comment]
Hi tiangolo,

if I understand you correctly, you're saying that if I add another layer that collect all the tasks of upload file and then run read  on each file asynchronically?
I can tell you that I debugged the code and I saw that upload2 is indeed called asynchronically, so I dont understand how adding another layer of collecting all the uploads tasks will help.
also I found the problem in this code:
`rand_file = generate_rand_data_file(data_size=file_size)
header = {'filename': os.path.basename(rand_file), 'content-length': str(os.path.getsize(rand_file))}
def gen_file_stream():
    with open(rand_file , "rb") as f:
        while True:
            data = f.read(1 * 1024 * 1024)
            if not data:
                break
            yield data
res = client.post('/upload', data=gen_file_stream)`

the line:
`res = client.post('/upload', data=gen_file_stream, headers=header)`` is wrong it needs to be:
`res = client.post('/upload', data=gen_file_stream())``
and now it works.

The thing is if I use requests package to post that upload request, that chunk remains 64k even though Im trying to upload 1MB.
My effort now is to try and increase the size of the chunk, I would be happy to know if somebody succeeded to do it with requests package

**Comment by tiangolo** [comment]
Ah, cool, if you solved your problem on the FastAPI side, then you could close the issue, the part that is specific about Requests would probably be better asked in StackOverflow or similar.
